packages/ai-jsx/src/lib/asr/asr.tsx

```diff
@@ -0,0 +1,428 @@
+    this.outBuf = [];
+    this.context = new window.AudioContext(); 
+    this.stream = stream;
+    const workletSrcBlob = new Blob([AUDIO_WORKLET_SRC], {
+      type: 'application/javascript',
+    });
+    const workletSrcUrl = URL.createObjectURL(workletSrcBlob);
+    await this.context!.audioWorklet.addModule(workletSrcUrl);
+    this.processorNode = new AudioWorkletNode(this.context, 'input-processor');
+    this.processorNode.port.onmessage = (event) => {
+      this.outBuf!.push(event.data);
+      const bufDuration =
+        ((128 * this.outBuf!.length) / this.sampleRate()) * 1000;
+      if (bufDuration >= timeslice) {
+        const chunkEvent = new CustomEvent('chunk', {
+          detail: this.makePcmChunk(this.outBuf!),
+        });
+        this.dispatchEvent(chunkEvent);
+        this.outBuf = [];
+      }
+    };
+    const source = this.context.createMediaStreamSource(stream);
+    source.connect(this.processorNode);
+    this.numSilentFrames = 0;
+  }
+  stop() {
+    this.processorNode!.disconnect();
+    this.stream!.getTracks().forEach((track) => track.stop());
+    if (this.streamElement) {
+      this.streamElement.pause();
+    }
+    this.context!.close();
+    this.processorNode = undefined;
+    this.stream = undefined;
+    this.context = undefined;
+    this.outBuf = [];
+  }
+  sampleRate() {
+    return this.context?.sampleRate;
+  }
+  private makePcmChunk(buffers: Float32Array[]) {
+    const byteLength = buffers.reduce((sum, buf) => sum + buf.length, 0) * 2;
+    const pcmBuffer = new ArrayBuffer(byteLength);
+    const view = new DataView(pcmBuffer);
+    let index = 0;
+    let energy = 0.0;
+    buffers.forEach((buffer) => {
+      buffer.forEach((sample) => {
+        energy += sample * sample;
+        const i16 = Math.max(
+          -32768,
+          Math.min(32767, Math.floor(sample * 32768)),
+        );
+        view.setInt16(index, i16, true);
+        index += 2;
+      });
+      this.updateVad(energy / (index / 2));
+    });
+    return pcmBuffer;
+  }
+  private updateVad(energy: number) {
+    const dbfs = 10 * Math.log10(energy);
+    if (dbfs < -50) {
+      this.numSilentFrames++;
+      if (this.numSilentFrames == 100) {
+        this.dispatchEvent(new CustomEvent('vad', { detail: false }));
+      }
+    } else {
+      if (this.numSilentFrames >= 100) {
+        this.dispatchEvent(new CustomEvent('vad', { detail: true }));
+      }
+      this.numSilentFrames = 0;
+    }
+  }
+}
+
+export class SpeechRecognitionBase extends EventTargetImpl {
+  private lastUtteranceEndTime: number;
+  protected manager: MicManager;
+  protected socket?: WebSocket;
+  protected name: string;
+  protected language?: string;
+  private outBuf: ArrayBuffer[];
+  constructor(manager: MicManager, name: string, language?: string) {
+    super();
+    this.manager = manager;
+    this.name = name;
+    this.language = language;
+    this.lastUtteranceEndTime = 0;
+    this.outBuf = [];
+  }
```

Comment:
```suggestion
export class SpeechRecognitionBase extends EventTargetImpl {
  private lastUtteranceEndTime: number = 0;
  private outBuf: ArrayBuffer[] = [];
  protected socket?: WebSocket;

  constructor(
    protected manager: MicManager,
    protected name: string,
    protected language?: string
  ) {
    super();
  }
}
```

This can be shorter via https://www.typescriptlang.org/docs/handbook/classes.html#parameter-properties.